Adding possibility to specify due date and time with natural language

DomizianoScarcelli opened the pull request:
**Note: This is my first contribution using Swift. I've never used Swift and SwiftUI until now, so I apologize if the quality of the code is not the best.** 

I love this app, the only thing missing is the possibility to add the reminder due date and time by simply writing it in natural language, as it already happens in the Apple Reminders app. This has been cited here (https://github.com/DamascenoRafael/reminders-menubar/issues/17) and here (https://github.com/DamascenoRafael/reminders-menubar/issues/104).

I used the [SwiftyChrono](https://github.com/quire-io/SwiftyChrono) package to achieve this without having to write a lot of code to create the language parser from scratch. The package isn't maintained anymore, but for a temporary solution, it gets the job done. 

![ezgif-1-d51af9f513](https://user-images.githubusercontent.com/44399141/218276626-5aa7c1a9-3946-45aa-a9e6-c74434f73c7e.gif)

Unfortunately, the supported languages in this package today are a few (English, French, Japanese, Spanish, and Chinese), and they do not cover all the languages supported in the current project.

I'm currently opening this pull request to let you know that I'm working on this feature. For now, in English it performs decently, still, there may be some bugs here and there.

Things I'm working on:
- [x] SwiftLint syntax check
- [x] App crashes when parsing strings like "Task on 12 February" (The bug was in the SwiftyChrono package, so I had to include the Pods folder in the tracked files, otherwise the code would be modified only at my end)
- [x] Date-related words have to be removed from the reminder title
- [x] The language changes only when the app is restarted, this has to be solved for the parser language to change every time the app language is modified;
- [x] When a time-related keyword (i.e. "at 9pm") is deleted, the shown time is reset, on the other hand, this doesn't happen for date-related keywords (i.e. "tomorrow").

Things that can be implemented in the future:
- Allow the user to deactivate the functionality in the menu
- Allow parsing for more languages, by updating the SwiftyChrono code inside the Pods folder (since I don't think the package is maintained anymore)
- Since SwiftyChrono isn't maintained anymore, this has to be considered as a temporary solution, and it may be possible to embed the date parser directly in the project without using external packages.
- SwiftyChrono doesn't recognize dates in some patterns like "Task on February the 12th".

DamascenoRafael commented:
@DomizianoScarcelli thank you for putting effort into this 🎉 
I'll be trying to review and test this over the weekend.

DamascenoRafael commented on .gitignore:
```diff
@@ -54,7 +54,7 @@ playground.xcworkspace
 # https://guides.cocoapods.org/using/using-cocoapods.html#should-i-check-the-pods-directory-into-source-control
 #
-# Pods/
+#Pods/
```
Please remove the Pods folder from git. After that remove the "#" from this line so that the folder will be ignored in the future.

DomizianoScarcelli replied:
Yes, I wanted to do this, but I had to modify a line in the SwiftyChrono package because the German language date parser causes an error that crashes everything, and I don't know how to solve it without changing the actual code in the package.

DomizianoScarcelli replied:
Nevermind, I solved it 👍🏽

DamascenoRafael commented on reminders-menu-bar/Views/FormNewReminderView.swift:
```diff
@@ -53,6 +54,16 @@ struct FormNewReminderView: View {
         .onChange(of: rmbReminder.title) { newValue in
             withAnimation(.easeOut(duration: 0.3)) {
                 isShowingDueDateOptions = !newValue.isEmpty
+                if !dateParser.isLanguageSupported {return}
+                guard let (parsedDate, dateRelatedText) = dateParser.buildDate(from: newValue) else {
+                    rmbReminder.hasTime = false
+                    rmbReminder.hasDueDate = false
+                    return
+                }
+                rmbReminder.dateRelatedText = dateRelatedText
+                rmbReminder.hasTime = dateParser.isTimeDefined
+                rmbReminder.hasDueDate = dateParser.isDateDefined
+                rmbReminder.date = parsedDate
```
I think we can move this logic into the RmbReminder class, maybe in the title didSet.

DomizianoScarcelli commented:
@DamascenoRafael I refactored the code a little bit because the main function `buildDate` of `NLPDateParser` was quite a mess, I tested it on every case that comes to my mind and it seems to work perfectly.

I will also write here the logic that I implemented, hoping it will be more clear for you to review and test the feature:
- If the user only specifies the time, and this is after the current time then we assume it's for today; otherwise we assume it's for the next day
- If the user only specifies the date, then the time is not defined. 
- If the user only specifies the day and month, then if it's later in the year, we assume it's for the same year; otherwise we assume it's for the next year.
- If the user specifies the full date, and this is in the past, then it's not considered valid

### Examples:
Let's say the current date is 17 Feb 2023 - 4:56PM
- "Task at 9pm" $\rightarrow$ 17 Feb 2023 - 9:00PM
- "Task at 9am" $\rightarrow$ 18 Feb 2023 - 9:00AM
- "Task 13 mar" $\rightarrow$ 13 March 2023 - No time
- "Task 20 January" $\rightarrow$ 20 January 2024 - No time
- "Task 1 January 2023" $\rightarrow$ No Date - No Time

DamascenoRafael commented:
@DomizianoScarcelli I did some search today about this feature and how it could be implemented.

I found that there is an Apple API (_NSDataDetector_) that can accomplish this.

The main advantages that I can see in this API are:
/- No need to use CocoaPods (or a lib that isn't maintained anymore)
/- Supports multiple languages (apparently without having to specify which one)

Let me know if you see any downsides to this approach.

I did some tests in Swift Playground and built the class below:
```swift
import Foundation

class DateParser {
    static let instance = DateParser()
    
    private let detector: NSDataDetector?
    
    struct DateParseResult {
        let date: Date
        let hasTime: Bool
    }
    
    private init() {
        // This prevents others from using the default '()' initializer for this class.
        let types: NSTextCheckingResult.CheckingType = [.date]
        detector = try? NSDataDetector(types: types.rawValue)
    }
    
    private func adjustDateAccordingToNow(_ dateResult: DateParseResult) -> DateParseResult {
        // TODO: **Adjustments you mentioned depending on whether the date is in the past.**
        return dateResult
    }
    
    func getDate(from textString: String) -> DateParseResult? {
        let range = NSRange(textString.startIndex..., in: textString)
        
        let matches = detector?.matches(in: textString, options: [], range: range)
        guard let match = matches?.first, let date = match.date else {
            return nil
        }
        
        var hasTime = false
        let timeIsSignificantKey = "timeIsSignificant"
        if match.responds(to: NSSelectorFromString(timeIsSignificantKey)) {
            hasTime = match.value(forKey: timeIsSignificantKey) as? Bool ?? false
        }
        
        let dateResult = DateParseResult(date: date, hasTime: hasTime)
        return adjustDateAccordingToNow(dateResult)
    }
}
```

Here are some results I got:
<img width="931" alt="image" src="https://user-images.githubusercontent.com/9842644/219981061-23e0f980-901c-4e1f-9e30-02011a9b776b.png">

Sorry for only finding this API now, I had never heard of it.
Let me know what you think. If you want you can try to add this class instead of SwiftyChrono.

DomizianoScarcelli commented:
Oh ok, I wasn't aware of this package. Yes, I think it would absolutely be better than using SwiftyChrono. I went with it because it seemed to be the only "fast" way to implement the feature. These days I'll try to implement it with the Apple API.

DomizianoScarcelli commented:
@DamascenoRafael I changed the date parser to use NSDataDetector and it works fine! Amazingly, it works for every language without even having to specify it. The only problem is that it doesn't recognize patterns like "In 10 minutes" or others like that. So it's a little bit of a pain to have to specify the exact hour each time. But at least it's a great trade-off considering the full language support.
Later I will do some more research on how to implement this missing feature, and if it's doable or not.

DamascenoRafael commented on reminders-menu-bar.xcodeproj/project.pbxproj:
```diff
@@ -582,7 +590,7 @@
 				DEAD_CODE_STRIPPING = YES;
 				DEVELOPMENT_ASSET_PATHS = "";
-				DEVELOPMENT_TEAM = 3398VKLH28;
+				DEVELOPMENT_TEAM = DJ42YWF9UH;
```
Please revert the DEVELOPMENT_TEAM change to keep the previous one.

```suggestion
				DEVELOPMENT_TEAM = 3398VKLH28;
```

DamascenoRafael commented on reminders-menu-bar.xcodeproj/project.pbxproj:
```diff
@@ -608,7 +616,7 @@
 				DEAD_CODE_STRIPPING = YES;
 				DEVELOPMENT_ASSET_PATHS = "";
-				DEVELOPMENT_TEAM = 3398VKLH28;
+				DEVELOPMENT_TEAM = DJ42YWF9UH;
```
Please revert the DEVELOPMENT_TEAM change to keep the previous one.

```suggestion
				DEVELOPMENT_TEAM = 3398VKLH28;
```

DamascenoRafael commented on reminders-menu-bar.xcodeproj/project.pbxproj:
```diff
@@ -753,7 +761,7 @@
 				DEAD_CODE_STRIPPING = YES;
 				DEVELOPMENT_ASSET_PATHS = "\"reminders-menu-bar/Preview Content\"";
-				DEVELOPMENT_TEAM = 3398VKLH28;
+				DEVELOPMENT_TEAM = DJ42YWF9UH;
```
Please revert the DEVELOPMENT_TEAM change to keep the previous one.

```suggestion
				DEVELOPMENT_TEAM = 3398VKLH28;
```

DamascenoRafael commented on reminders-menu-bar.xcodeproj/project.pbxproj:
```diff
@@ -780,7 +788,7 @@
 				DEAD_CODE_STRIPPING = YES;
 				DEVELOPMENT_ASSET_PATHS = "\"reminders-menu-bar/Preview Content\"";
-				DEVELOPMENT_TEAM = 3398VKLH28;
+				DEVELOPMENT_TEAM = DJ42YWF9UH;
```
Please revert the DEVELOPMENT_TEAM change to keep the previous one.

```suggestion
				DEVELOPMENT_TEAM = 3398VKLH28;
```

DamascenoRafael commented on reminders-menu-bar/Models/DateParser.swift:
```diff
@@ -0,0 +1,89 @@
+            && inputDateComponents.year == todayDateComponents.year
+    }
+    
+    private func checkIfPastTime(from date: Date) -> Bool {
```
In `Date+Extensions.swift` we have an `isPast` variable which I believe does what you want to check here. In that case we can remove this `checkIfToday` function and use the existing `isPast` variable instead.

DamascenoRafael commented on reminders-menu-bar/Models/DateParser.swift:
```diff
@@ -0,0 +1,89 @@
+        return adjustDateAccordingToNow(dateResult)
+    }
+    
+    private func checkIfToday(from date: Date) -> Bool {
```
Please move this function to `Date+Extensions.swift`, but change it to a variable format like we have for `isPast`. Call this variable `var isToday: Bool`.

When calling `.dateComponents` to calculate whether the date is today please remove `.hour` and `.minute` from the array, as these components are not used to determine whether the date is today.

DamascenoRafael replied:
We can do the same thing for `func checkIfSameYear` and call it `var isThisYear: Bool`, since the comparison is always with the current year.

DamascenoRafael commented on reminders-menu-bar/Models/DateParser.swift:
```diff
@@ -0,0 +1,89 @@
+    private func adjustDateAccordingToNow(_ dateResult: DateParseResult) -> DateParseResult? {
+        // If the date it's not in the current year, then it's not valid
+        if !checkIfSameYear(from: dateResult.date) {
+            return nil
```
Can you give me more details on why we consider it invalid and return nil in this case?

In a situation where the user specifies a year in the future, like _"send paper work on 9 mar 2024"_ for example, I think we should keep the date in the future.

DomizianoScarcelli replied:
Yes that's my mistake, the idea is to discard the events in the past years

DamascenoRafael reviewed (commented):
hey @DomizianoScarcelli, thanks for the hard work on this feature!!
I've added some comments for improvements and questions about the code.
Let me know if you have any questions.

DomizianoScarcelli commented:
@DamascenoRafael Thanks for the review! I changed the things you mentioned 😄

DamascenoRafael commented:
Thanks @DomizianoScarcelli
I'm going to merge this feature, test it some more and if everything looks ok it will be in the next release 📦 
Thanks again for putting effort into this!! 🚀
